Spotlight UI skeleton

nickoferrall opened the pull request:
PR to resolve issue #5106

nickoferrall commented on packages/client/components/SpotlightModal.tsx:
```diff
@@ -0,0 +1,207 @@
+  const reflectionGroupsCount = 0
+  return (
+    <ModalContainer onKeyDown={handleKeyDown} isDesktop={isDesktop}>
+      <SelectedReflection>
```
Can move `SelectedReflection` and `SimilarReflectionGroups` to their own components once the logic is built up a bit more

nickoferrall commented on packages/client/components/ReflectionCard/ReflectionCard.tsx:
```diff
@@ -55,16 +83,29 @@ const getReadOnly = (
   const [editorState, setEditorState] = useEditorState(content)
-
+  const [isHovering, setIsHovering] = useState(false)
+  const isDesktop = useBreakpoint(Breakpoint.SIDEBAR_LEFT)
```
I find some of the breakpoints a bit confusing, e.g. why `SIDEBAR_LEFT` and `NEW_MEETING_SELECTOR` instead of say XS, S, M?

mattkrick replied:
fair point!
there are 2 schools of thought:
- bins by device size (S,M, L or maybe phone, tablet, desktop, TV)
- bins by functionality (hide the left nav, show 2 columns instead of 3, etc)

binning by device size was originally really popular, but becomes a headache if the sizes of devices vary widely. you're basically building your app to a set number of devices. it's still pretty popular for UI frameworks!

binning by functionality means the layout is determined by your components, not the hardware specs of your users. at least that's the thought.

that said, we've done a real bad job of sticking with 1 approach :see_no_evil:

nickoferrall replied:
Makes sense - appreciate the explanation!

enriquesanchez reviewed (changes requested):
This is looking amazing @nickoferrall 🙌

Here's my feedback:

1. The spotlight modal should have a dropshadow. Let's use the same value we use for other modals/popups: 

`box-shadow: rgba(0, 0, 0, 0.2) 0px 5px 5px -3px, rgba(0, 0, 0, 0.14) 0px 8px 10px 1px, rgba(0, 0, 0, 0.12) 0px 3px 14px 2px;`

2. Is there a way to vertically align the title and the `X` button?
<img width="766" alt="Screen Shot 2021-07-26 at 5 10 18 p m" src="https://user-images.githubusercontent.com/3276087/127065785-b5642a8a-6999-42e9-9408-52647c620628.png">

3. We are going to have some reflection cards that are small (one line of text), and others that are taller. Ideally we can vertically center these instead of having them aligned to the top. Would this be possible?

<img width="392" alt="Screen Shot 2021-07-26 at 5 11 21 p m" src="https://user-images.githubusercontent.com/3276087/127065923-6a4a02ad-2c36-4dbc-988c-261eb1f8f83c.png">

4. Any chance we can increase the color contrast of the search field's placeholder text? In the mockups I'm using `Slate 600 #82809A`.

nickoferrall commented:
Thanks for the review! 

I've implemented those changes. I was umming and arring about aligning the card at the top or center but I think you're right about center alignment. The placeholder text in that screenshot is lighter than how it looks for me - I've added `Slate 600` which looks similar to how it was before so please let me know if it needs to be any darker 👍 

As mentioned in the description, I'm working on fixing the TypeScript error. I've noticed that when you click the spotlight on a reflection and close the modal, the search icon no longer appears for that reflection. To fix, I think I need to either clone the reflection or move the position of that reflection to the modal. I'll look into how the reflection stack logic works in more detail. Either way, it's ready for a re-review @enriquesanchez when you get the chance

enriquesanchez commented:
@nickoferrall Thanks! This is looking :chef_kiss:

>I was umming and arring about aligning the card at the top or center but I think you're right about center alignment. 

I also like the center alignment. Looks great on big and small viewports:

![image](https://user-images.githubusercontent.com/3276087/127227114-959678c6-2ad9-4077-8097-380631c93ba4.png)

![image](https://user-images.githubusercontent.com/3276087/127227164-fdb32853-32a0-4e61-91ab-a22685409c0e.png)

>The placeholder text in that screenshot is lighter than how it looks for me

Ah, this is a Firefox thing. By default it lightens the placeholder text. Tested on Chrome and the color contrast is great.

nickoferrall commented on packages/client/components/SpotlightModal.tsx:
```diff
@@ -0,0 +1,214 @@
+            isReadOnly
+            staticIdx={0}
+            reflection={reflection as any}
+            meeting={meeting as any}
```
I've been having problems getting the right types for the meeting and reflection. It looks like we may have had a similar issue in [ExpandedReflection](https://github.com/ParabolInc/parabol/blob/8c68bf771a350b31c7948e6bd11c5c110e30c977/packages/client/components/RetroReflectPhase/ExpandedReflection.tsx#L15). Ideally, I'd like to just query the bare minimum fields we need in the file, the reflection & meeting ids, but I'm having difficulty as TypeScript complains about missing refs. I think this is because there's a circular reference: the reflection card contains the spotlight modal which contains the reflection card

nickoferrall commented on packages/client/components/ReflectionGroup/ReflectionGroup.tsx:
```diff
@@ -116,8 +116,10 @@ const ReflectionGroup = (props: Props) => {
-    const wasDrag = staticReflections.some((reflection) => reflection.isDropping)
-    if (wasDrag) return
+    const isNotDraggable = staticReflections.some(
+      (reflection) => reflection.isDropping || reflection.inSpotlight
```
If you have a stack of reflections, click the `showSpotlight` icon and close the spotlight modal, the reflection stack modal will then open up when we don't want it to. Using `stopPropagation` when clicking the `showSpotlight` icon doesn't prevent it from happening so it looks like we need to prevent it in the `onClick` here. 

To do this, I've added an `inSpotlight` field to the Relay client. This seems like overkill & is confusing as we also need to pass an `inSpotlight` prop to the `draggableReflection` in the modal so the `showSpotlight` icon doesn't show (Relay doesn't update in time to work in both). I'll work on finding a better solution but any suggestions would be great!

mattkrick commented on packages/client/components/ReflectionCard/ReflectionCard.tsx:
```diff
@@ -179,41 +224,89 @@ const ReflectionCard = (props: Props) => {
+        </SearchButton>
+        {tooltipPortal('Find similar')}
+      </ReflectionCardRoot>
+      {modalPortal(
```
let's move this logic to the RetroGroupPhase or GroupingKanban. then we can pass down something like:
```ts
const openSpotlight = (reflection: Reflection) => {
  commitLocalUpdate(atmosphere, (store) => store.get(meetingId).setLinkedRecord(reflection, 'spotlightReflection')
})
```
then we only open the modal if it was falsy & now it's a reflection object.
that'll express that there can only be 1 spotlight open & a spotlight takes in a reflection.

mattkrick commented on packages/client/components/ReflectionGroup/DraggableReflectionCard.tsx:
```diff
@@ -88,7 +92,7 @@ const DraggableReflectionCard = (props: Props) => {
     swipeColumn
   )
-  const isDragPhase = phaseType === 'group' && !isComplete
+  const isDragPhase = phaseType === 'group' && !isComplete && !inSpotlight
```
we'll want the cards to be draggable inside the spotlight & if someone else is dragging the card it should be moving around in the spotlight, too

nickoferrall replied:
Makes sense. Just to make sure we're on the same page with how the spotlight dragging behaviour will work: 

- if the viewer opens the spotlight modal, the spotlight modal only opens for the viewer, not for everyone in the meeting
- if someone that is not the viewer drags the reflection that's in the spotlight, the reflection would drag as normal for the non-viewer
- if someone that is not the viewer drags the reflection that's in the spotlight into a new reflection group, nothing would change in the spotlight for the viewer. If the viewer then drags the spotlight reflection into one of the reflection groups in the spotlight, the reflection would move to the new reflection group
- if the viewer moves the spotlight reflection to a reflection group in the spotlight, non-viewers would see the remote reflection move from it's original position on the kanban to its new destination

mattkrick replied:
> if someone that is not the viewer drags the reflection that's in the spotlight into a new reflection group, nothing would change in the spotlight for the viewer.

the opposite for this one. if you've got spotlight open and see the card "Cats", I can drag that around. if i hover a group in your spotlight, it you see that. if i hover a group not in your spotlight, you just see it floating somewhere. if i release the mouse, you see the card go back to its spot. if i drop it on a group in your spotlight, you see that group grow in size.

nickoferrall commented:
@mattkrick as #5271 builds on top of the work here (and makes a few small changes to the code in this PR) would you be happy for me to close this PR? Once Enrique approves #5271, you could just review the code in that PR instead of both this PR and that one?

mattkrick commented:
sounds like a good plan!
